### ISSUE TITLE
Support for masking (variable length input)

### ISSUE THREAD

**Issue by sensout** [Element-Research/rnn#45]
Thank you for this great RNN implementation.

I am trying to batch input data with a variable number of timesteps through an LSTM.
Is there a simple way to support this feature with your implementation ?

I am thinking about something like the `mask_zero` in http://keras.io/layers/recurrent/ and http://keras.io/layers/embeddings/ where the embedding would be a `nn.LookupTable` always returning a zero norm vector for padding.

Thanks

**Comment by nicholas-leonard** [comment]
Hi @sensout . We don't have anything like `mask_zero`. However, for variable length sequences, I like to group sequences into groups of the same length. Then each batch can consist of sequences of the same length. But different batches can have different lengths. The RNN package supports that (variable length batches, not variable length examples). 

On a different note, I don't understand how the keras package uses mask_zero. So it would seem that, within a batch, the smaller sequences are padded with zeros to match the length of the longer one. Either those zero-embeddings are propagate through the rnn just like any other embedding, or the rnn can somehow ignore them. In the first case, I think it would make most sense to pad the first elements instead of the last, but the rnn would still be accumulated bias, so that wouldn't work. But in the second case, I guess you could still pad the first, but detect a zero embedding (min() - max() = 0) and just zero the hidden activation. That would be reasonably easy to implement. 

Ideally, in order to reduce code bloat, it would be possible to implement this feature as its own module, which we can include or not into AbstractRecurrent instances calling a method like : maskZero(). So I am thinking the module could decorate the internal recurrentModule, which takes the current input and previous hidden state as input. For each row (i.e. example) in the current input state, if its a tensor of zeros, then the corresponding row in the output state would also be zeroed. This module could be called MaskZero, it would inherit the nn.Decorator module. The gradInputs would also be zeros. In any case, the entire batch would be propagated, unless a faster way can be implemented using torch.indexCopy or torch.index. 

Would you like to give it a shot and send a PR? I could help you with the details.

**Comment by sensout** [comment]
Thank you for your quick answer.

Grouping sequences of same length does help performance but unless you accept to drop sequences which do not belong to any batch, you still need padding.

For now I am grouping sequences of same length and padding the few ones which do not fit any batch perfectly. I secretly hope that the LSTM will learn to ignore these paddings (forget them) but this is far from efficient (nor obvious).

I agree with you about the bias problem for zero-embeddings. 
Zeroing the hidden activation seems the way to go.

Thank you for pointing the right place to implement this. 
I will try to make a PR this week.

Cheers

**Comment by nicholas-leonard** [comment]
Awesome, and don't hesitate to PR anything you have. I don't care if its a draft, at least it would be something for me to go on. And I love to see people other than me in the contributors list!

**Comment by sensout** [comment]
Great !
And forget my remark about dropping sequences if batches can have different lengths. 
I just realized what you meant.
